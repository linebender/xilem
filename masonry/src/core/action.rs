// Copyright 2022 the Xilem Authors
// SPDX-License-Identifier: Apache-2.0

use std::any::Any;

use crate::core::PointerButton;

// TODO - Replace actions with an associated type on the Widget trait
// See https://github.com/linebender/xilem/issues/664

// TODO - TextCursor changed, ImeChanged, EnterKey, MouseEnter
#[non_exhaustive]
/// Events from UI elements.
///
/// Note: Actions are still a WIP feature.
pub enum Action {
    /// A button was pressed.
    ButtonPressed(PointerButton),
    /// Text changed.
    TextChanged(String),
    /// Text entered.
    TextEntered(String),
    /// A checkbox was toggled.
    CheckboxToggled(bool),
    // FIXME - This is a huge hack
    /// Other.
    Other(Box<dyn Any + Send>),
}

impl PartialEq for Action {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::ButtonPressed(l_button), Self::ButtonPressed(r_button)) => l_button == r_button,
            (Self::TextChanged(l0), Self::TextChanged(r0)) => l0 == r0,
            (Self::TextEntered(l0), Self::TextEntered(r0)) => l0 == r0,
            (Self::CheckboxToggled(l0), Self::CheckboxToggled(r0)) => l0 == r0,
            // FIXME
            // (Self::Other(val_l), Self::Other(val_r)) => false,
            _ => false,
        }
    }
}

impl std::fmt::Debug for Action {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ButtonPressed(button) => f.debug_tuple("ButtonPressed").field(button).finish(),
            Self::TextChanged(text) => f.debug_tuple("TextChanged").field(text).finish(),
            Self::TextEntered(text) => f.debug_tuple("TextEntered").field(text).finish(),
            Self::CheckboxToggled(b) => f.debug_tuple("CheckboxChecked").field(b).finish(),
            Self::Other(_) => write!(f, "Other(...)"),
        }
    }
}
