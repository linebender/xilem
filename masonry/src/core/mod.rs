// Copyright 2025 the Xilem Authors
// SPDX-License-Identifier: Apache-2.0

//! Basic types and traits Masonry is built on.

mod action;
mod box_constraints;
mod contexts;
mod event;
mod object_fit;
mod text;
#[allow(missing_docs, reason = "TODO")]
mod widget;
mod widget_arena;
#[allow(missing_docs, reason = "TODO")]
mod widget_mut;
#[allow(missing_docs, reason = "TODO")]
mod widget_pod;
mod widget_ref;
mod widget_state;

pub use action::Action;
pub use box_constraints::BoxConstraints;
pub use contexts::{
    AccessCtx, ComposeCtx, EventCtx, IsContext, LayoutCtx, MutateCtx, PaintCtx, QueryCtx,
    RawWrapper, RawWrapperMut, RegisterCtx, UpdateCtx,
};
pub use event::{
    AccessEvent, PointerButton, PointerEvent, PointerState, TextEvent, Update, WindowEvent,
    WindowTheme,
};
pub use object_fit::ObjectFit;
pub use text::{render_text, ArcStr, BrushIndex, StyleProperty, StyleSet};
pub use widget::find_widget_at_pos;
pub use widget::{AllowRawMut, FromDynWidget, Widget, WidgetId};
pub use widget_mut::WidgetMut;
pub use widget_pod::WidgetPod;
pub use widget_ref::WidgetRef;

pub(crate) use text::default_styles;
pub(crate) use widget_arena::WidgetArena;
pub(crate) use widget_pod::CreateWidget;
pub(crate) use widget_state::WidgetState;
