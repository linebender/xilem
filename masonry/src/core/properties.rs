// Copyright 2025 the Xilem Authors
// SPDX-License-Identifier: Apache-2.0

use anymap3::{AnyMap, Entry};

// TODO - Add PropertyValue wrapper struct that implements receiver trait.
// Return PropertyValue<T> instead of Option<T> from methods.

/// A collection of properties that a widget can be created with.
///
/// See [properties documentation](crate::doc::doc_04b_widget_properties) for details.
#[derive(Default)]
pub struct Properties {
    pub(crate) map: AnyMap,
}

/// Reference to a collection of properties that a widget has access to.
///
/// Used by the [`Widget`] trait during rendering passes and in some search methods.
///
/// See [properties documentation](crate::doc::doc_04b_widget_properties) for
/// details.
///
/// [`Widget`]: crate::core::Widget
#[derive(Clone, Copy)]
pub struct PropertiesRef<'a> {
    pub(crate) map: &'a AnyMap,
}

/// Mutable reference to a collection of properties that a widget has access to.
///
/// Used by the [`Widget`] trait during most passes.
///
/// See [properties documentation](crate::doc::doc_04b_widget_properties) for
/// details.
///
/// [`Widget`]: crate::core::Widget
pub struct PropertiesMut<'a> {
    pub(crate) map: &'a mut AnyMap,
}

impl Properties {
    /// Create an empty collection of properties.
    pub fn new() -> Self {
        Self { map: AnyMap::new() }
    }

    /// Get a reference to the properties.
    pub fn ref_(&self) -> PropertiesRef<'_> {
        PropertiesRef { map: &self.map }
    }

    /// Get a mutable reference to the properties.
    pub fn mut_(&mut self) -> PropertiesMut<'_> {
        PropertiesMut { map: &mut self.map }
    }
}

impl PropertiesRef<'_> {
    /// Returns true if the widget has a property of type `T`.
    pub fn contains<T: 'static>(&self) -> bool {
        self.map.contains::<T>()
    }

    /// Get value of property `T`, or None if the widget has no `T` property.
    pub fn get<T: 'static>(&self) -> Option<&T> {
        self.map.get::<T>()
    }
}

impl PropertiesMut<'_> {
    /// Returns true if the widget has a property of type `T`.
    pub fn contains<T: 'static>(&self) -> bool {
        self.map.contains::<T>()
    }

    /// Get value of property `T`, or None if the widget has no `T` property.
    pub fn get<T: 'static>(&self) -> Option<&T> {
        self.map.get::<T>()
    }

    /// Get value of property `T`, or None if the widget has no `T` property.
    ///
    /// If you're using a `WidgetMut`, call [`WidgetMut::get_prop_mut`] instead.
    ///
    /// [`WidgetMut::get_prop_mut`]: crate::core::WidgetMut::get_prop_mut
    pub fn get_mut<T: 'static>(&mut self) -> Option<&mut T> {
        self.map.get_mut::<T>()
    }

    /// Set property `T` to given value. Returns the previous value if `T` was already set.
    ///
    /// If you're using a `WidgetMut`, call [`WidgetMut::insert_prop`] instead.
    ///
    /// [`WidgetMut::insert_prop`]: crate::core::WidgetMut::insert_prop
    pub fn insert<T: 'static>(&mut self, value: T) -> Option<T> {
        self.map.insert(value)
    }

    /// Remove property `T`. Returns the previous value if `T` was set.
    ///
    /// If you're using a `WidgetMut`, call [`WidgetMut::remove_prop`] instead.
    ///
    /// [`WidgetMut::remove_prop`]: crate::core::WidgetMut::remove_prop
    pub fn remove<T: 'static>(&mut self) -> Option<T> {
        self.map.remove::<T>()
    }

    /// Returns an entry that can be used to add, update, or remove a property.
    ///
    /// If you're using a `WidgetMut`, call [`WidgetMut::prop_entry`] instead.
    ///
    /// [`WidgetMut::prop_entry`]: crate::core::WidgetMut::prop_entry
    pub fn entry<T: 'static>(&mut self) -> Entry<'_, dyn std::any::Any, T> {
        self.map.entry::<T>()
    }

    /// Get a `PropertiesMut` for the same underlying properties with a shorter lifetime.
    pub fn reborrow_mut(&mut self) -> PropertiesMut<'_> {
        PropertiesMut {
            map: &mut *self.map,
        }
    }
}
