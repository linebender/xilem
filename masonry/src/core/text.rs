// Copyright 2018 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

//! Support for text display and rendering
//!
//! There are three kinds of text commonly needed:
//!  1) Non interactive text (e.g. a button's label)
//!  2) Selectable text (e.g. a paragraph of content)
//!  3) Editable text (e.g. a search bar)
//!
//! All of these have the same set of global styling options, and can contain rich text

use parley::GenericFamily;

/// A reference counted string slice.
///
/// This is a data-friendly way to represent strings in Masonry. Unlike `String`
/// it cannot be mutated, but unlike `String` it can be cheaply cloned.
pub type ArcStr = std::sync::Arc<str>;

/// The Parley [`Brush`] used within Masonry.
///
/// This enables updating of brush details without performing relayouts;
/// the inner values are indexes into the `brushes` argument to [`render_text()`].
///
/// [`Brush`]: crate::parley::Brush
#[derive(Clone, PartialEq, Default, Debug)]
pub struct BrushIndex(pub usize);

/// A style property specialised for use within Masonry.
pub type StyleProperty = parley::StyleProperty<'static, BrushIndex>;

/// A set of styles specialised for use within Masonry.
pub type StyleSet = parley::StyleSet<BrushIndex>;

/// Applies the default text styles for Masonry into `styles`.
pub(crate) fn default_styles(styles: &mut StyleSet) {
    styles.insert(StyleProperty::LineHeight(1.2));
    styles.insert(GenericFamily::SystemUi.into());
}

use parley::{Layout, PositionedLayoutItem};
use vello::kurbo::{Affine, Line, Stroke};
use vello::peniko::{Brush, Fill};
use vello::Scene;

/// A function that renders laid out glyphs to a [`Scene`].
///
/// The `BrushIndex` values of the runs are indices into `brushes`.
pub fn render_text(
    scene: &mut Scene,
    transform: Affine,
    layout: &Layout<BrushIndex>,
    brushes: &[Brush],
    // TODO: Should this be part of `BrushIndex` (i.e. `brushes`)?
    hint: bool,
) {
    for line in layout.lines() {
        for item in line.items() {
            let PositionedLayoutItem::GlyphRun(glyph_run) = item else {
                continue;
            };
            let style = glyph_run.style();
            // We draw underlines under the text, then the strikethrough on top, following:
            // https://drafts.csswg.org/css-text-decor/#painting-order
            if let Some(underline) = &style.underline {
                let underline_brush = &brushes[underline.brush.0];
                let run_metrics = glyph_run.run().metrics();
                let offset = match underline.offset {
                    Some(offset) => offset,
                    None => run_metrics.underline_offset,
                };
                let width = match underline.size {
                    Some(size) => size,
                    None => run_metrics.underline_size,
                };
                // The `offset` is the distance from the baseline to the top of the underline
                // so we move the line down by half the width
                // Remember that we are using a y-down coordinate system
                // If there's a custom width, because this is an underline, we want the custom
                // width to go down from the default expectation
                let y = glyph_run.baseline() - offset + width / 2.;

                let line = Line::new(
                    (glyph_run.offset() as f64, y as f64),
                    ((glyph_run.offset() + glyph_run.advance()) as f64, y as f64),
                );
                scene.stroke(
                    &Stroke::new(width.into()),
                    transform,
                    underline_brush,
                    None,
                    &line,
                );
            }
            let mut x = glyph_run.offset();
            let y = glyph_run.baseline();
            let run = glyph_run.run();
            let font = run.font();
            let font_size = run.font_size();
            let synthesis = run.synthesis();
            let glyph_xform = synthesis
                .skew()
                .map(|angle| Affine::skew(angle.to_radians().tan() as f64, 0.0));
            let coords = run.normalized_coords();
            let brush = &brushes[style.brush.0];
            scene
                .draw_glyphs(font)
                .brush(brush)
                .hint(hint)
                .transform(transform)
                .glyph_transform(glyph_xform)
                .font_size(font_size)
                .normalized_coords(coords)
                .draw(
                    Fill::NonZero,
                    glyph_run.glyphs().map(|glyph| {
                        let gx = x + glyph.x;
                        let gy = y - glyph.y;
                        x += glyph.advance;
                        vello::Glyph {
                            id: glyph.id as _,
                            x: gx,
                            y: gy,
                        }
                    }),
                );

            if let Some(strikethrough) = &style.strikethrough {
                let strikethrough_brush = &brushes[strikethrough.brush.0];
                let run_metrics = glyph_run.run().metrics();
                let offset = match strikethrough.offset {
                    Some(offset) => offset,
                    None => run_metrics.strikethrough_offset,
                };
                let width = match strikethrough.size {
                    Some(size) => size,
                    None => run_metrics.strikethrough_size,
                };
                // The `offset` is the distance from the baseline to the *top* of the strikethrough
                // so we calculate the middle y-position of the strikethrough based on the font's
                // standard strikethrough width.
                // Remember that we are using a y-down coordinate system
                let y = glyph_run.baseline() - offset + run_metrics.strikethrough_size / 2.;

                let line = Line::new(
                    (glyph_run.offset() as f64, y as f64),
                    ((glyph_run.offset() + glyph_run.advance()) as f64, y as f64),
                );
                scene.stroke(
                    &Stroke::new(width.into()),
                    transform,
                    strikethrough_brush,
                    None,
                    &line,
                );
            }
        }
    }
}
