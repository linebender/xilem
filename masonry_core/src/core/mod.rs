// Copyright 2025 the Xilem Authors
// SPDX-License-Identifier: Apache-2.0

//! Basic types and traits Masonry is built on.

mod action;
mod box_constraints;
mod contexts;
mod events;
mod object_fit;
mod properties;
mod text;
mod widget;
mod widget_arena;
mod widget_mut;
mod widget_pod;
mod widget_ref;
mod widget_state;

pub use action::Action;
pub use box_constraints::BoxConstraints;
pub use contexts::{
    AccessCtx, ComposeCtx, EventCtx, IsContext, LayoutCtx, MutateCtx, PaintCtx, QueryCtx,
    RawWrapper, RawWrapperMut, RegisterCtx, UpdateCtx,
};
pub use events::{
    AccessEvent, Force, Ime, PointerButton, PointerEvent, PointerState, ResizeDirection, TextEvent,
    Update, WindowEvent, WindowTheme,
};
pub use object_fit::ObjectFit;
pub use properties::{Properties, PropertiesMut, PropertiesRef};
pub use text::{ArcStr, BrushIndex, StyleProperty, StyleSet, render_text};
pub use widget::find_widget_under_pointer;
pub use widget::{AllowRawMut, FromDynWidget, Widget, WidgetId};
pub use widget_mut::WidgetMut;
pub use widget_pod::WidgetPod;
pub use widget_ref::WidgetRef;

pub(crate) use text::default_styles;
pub(crate) use widget_arena::WidgetArena;
pub(crate) use widget_pod::CreateWidget;
pub(crate) use widget_state::WidgetState;
