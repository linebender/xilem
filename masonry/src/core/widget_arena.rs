// Copyright 2024 the Xilem Authors
// SPDX-License-Identifier: Apache-2.0

use anymap3::AnyMap;
use tree_arena::{ArenaMut, ArenaRef, TreeArena};

use crate::core::{Widget, WidgetId, WidgetState};

pub(crate) struct WidgetArena {
    pub(crate) widgets: TreeArena<Box<dyn Widget>>,
    pub(crate) states: TreeArena<WidgetState>,
    pub(crate) properties: TreeArena<AnyMap>,
}

impl WidgetArena {
    pub(crate) fn has(&self, widget_id: WidgetId) -> bool {
        self.widgets.find(widget_id).is_some()
    }

    #[track_caller]
    pub(crate) fn parent_of(&self, widget_id: WidgetId) -> Option<WidgetId> {
        let widget_ref = self
            .widgets
            .find(widget_id)
            .expect("parent_of: widget not found in arena");

        let id = widget_ref.parent_id?;
        Some(WidgetId(id.try_into().unwrap()))
    }

    #[track_caller]
    pub(crate) fn get_all(
        &self,
        widget_id: WidgetId,
    ) -> (
        ArenaRef<Box<dyn Widget>>,
        ArenaRef<WidgetState>,
        ArenaRef<AnyMap>,
    ) {
        let widget = self
            .widgets
            .find(widget_id)
            .expect("get_pair: widget not in widget tree");
        let state = self
            .states
            .find(widget_id)
            .expect("get_pair: widget state not in widget tree");
        let properties = self
            .properties
            .find(widget_id)
            .expect("get_pair: widget properties not in widget tree");
        (widget, state, properties)
    }

    #[track_caller]
    pub(crate) fn get_all_mut(
        &mut self,
        widget_id: WidgetId,
    ) -> (
        ArenaMut<Box<dyn Widget>>,
        ArenaMut<WidgetState>,
        ArenaMut<AnyMap>,
    ) {
        let widget = self
            .widgets
            .find_mut(widget_id)
            .expect("get_pair_mut: widget not in widget tree");
        let state = self
            .states
            .find_mut(widget_id)
            .expect("get_pair_mut: widget state not in widget tree");
        let properties = self
            .properties
            .find_mut(widget_id)
            .expect("get_pair_mut: widget properties not in widget tree");
        (widget, state, properties)
    }

    #[allow(dead_code)]
    #[track_caller]
    pub(crate) fn get_widget(&self, widget_id: WidgetId) -> ArenaRef<Box<dyn Widget>> {
        self.widgets
            .find(widget_id)
            .expect("get_widget: widget not in widget tree")
    }

    #[allow(dead_code)]
    #[track_caller]
    pub(crate) fn get_widget_mut(&mut self, widget_id: WidgetId) -> ArenaMut<Box<dyn Widget>> {
        self.widgets
            .find_mut(widget_id)
            .expect("get_widget_mut: widget not in widget tree")
    }

    #[track_caller]
    pub(crate) fn get_state(&mut self, widget_id: WidgetId) -> ArenaRef<WidgetState> {
        self.states
            .find(widget_id)
            .expect("get_state: widget state not in widget tree")
    }

    #[track_caller]
    pub(crate) fn get_state_mut(&mut self, widget_id: WidgetId) -> ArenaMut<WidgetState> {
        self.states
            .find_mut(widget_id)
            .expect("get_state_mut: widget state not in widget tree")
    }
}
