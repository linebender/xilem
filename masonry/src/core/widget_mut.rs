// Copyright 2018 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

use std::any::TypeId;

use anymap3::Entry;

use crate::core::{FromDynWidget, MutateCtx, Widget};
use crate::kurbo::Affine;

// TODO - Document extension trait workaround.
// See https://xi.zulipchat.com/#narrow/stream/317477-masonry/topic/Thoughts.20on.20simplifying.20WidgetMut/near/436478885
/// A rich mutable reference to a [`Widget`].
///
/// In Masonry, widgets can't be mutated directly. All mutations go through a `WidgetMut`
/// wrapper. So, to change a label's text, you might call `WidgetMut<Label>::set_text()`.
/// This helps Masonry make sure that internal metadata is propagated after every widget
/// change.
///
/// You can create a `WidgetMut` from [`TestHarness`](crate::testing::TestHarness),
/// [`EventCtx`](crate::core::EventCtx), [`UpdateCtx`](crate::core::UpdateCtx) or from a parent
/// `WidgetMut` with [`MutateCtx`].
///
/// `WidgetMut` implements [`Deref`](std::ops::Deref) with `W::Mut` as target.
///
/// ## `WidgetMut` as a Receiver
///
/// Once the Receiver trait is stabilized, `WidgetMut` will implement it so that custom
/// widgets in downstream crates can use `WidgetMut` as the receiver for inherent methods.
#[non_exhaustive]
pub struct WidgetMut<'a, W: Widget + ?Sized> {
    pub ctx: MutateCtx<'a>,
    pub widget: &'a mut W,
}

impl<W: Widget + ?Sized> Drop for WidgetMut<'_, W> {
    fn drop(&mut self) {
        // If this `WidgetMut` is a reborrow, a parent non-reborrow `WidgetMut`
        // still exists which will do the merge-up in `Drop`.
        if let Some(parent_widget_state) = self.ctx.parent_widget_state.take() {
            parent_widget_state.merge_up(self.ctx.widget_state);
        }
    }
}

impl<W: Widget + ?Sized> WidgetMut<'_, W> {
    /// Get a `WidgetMut` for the same underlying widget with a shorter lifetime.
    pub fn reborrow_mut(&mut self) -> WidgetMut<'_, W> {
        let widget = &mut self.widget;
        WidgetMut {
            ctx: self.ctx.reborrow_mut(),
            widget,
        }
    }

    /// Returns true if the widget has a property of type `T`.
    pub fn get_prop<T: 'static>(&self) -> Option<&T> {
        self.ctx.properties.get::<T>()
    }

    /// Get value of property `T`, or None if the widget has no `T` property.
    pub fn contains_prop<T: 'static>(&self) -> bool {
        self.ctx.properties.contains::<T>()
    }

    /// Get value of property `T`, or None if the widget has no `T` property.
    pub fn get_prop_mut<T: 'static>(&mut self) -> Option<&mut T> {
        self.widget
            .property_changed(&mut self.ctx.update_mut(), TypeId::of::<T>());
        self.ctx.properties.get_mut::<T>()
    }

    /// Set property `T` to given value. Returns the previous value if `T` was already set.
    pub fn insert_prop<T: 'static>(&mut self, value: T) -> Option<T> {
        self.widget
            .property_changed(&mut self.ctx.update_mut(), TypeId::of::<T>());
        self.ctx.properties.insert(value)
    }

    /// Remove property `T`. Returns the previous value if `T` was set.
    pub fn remove_prop<T: 'static>(&mut self) -> Option<T> {
        self.widget
            .property_changed(&mut self.ctx.update_mut(), TypeId::of::<T>());
        self.ctx.properties.remove::<T>()
    }

    /// Returns an entry that can be used to add, update, or remove a property.
    pub fn prop_entry<T: 'static>(&mut self) -> Entry<'_, dyn std::any::Any, T> {
        self.widget
            .property_changed(&mut self.ctx.update_mut(), TypeId::of::<T>());
        self.ctx.properties.entry::<T>()
    }

    /// Set the local transform of this widget.
    ///
    /// It behaves similarly as CSS transforms.
    pub fn set_transform(&mut self, transform: Affine) {
        self.ctx.set_transform(transform);
    }

    /// Attempt to downcast to `WidgetMut` of concrete Widget type.
    pub fn try_downcast<W2: Widget + FromDynWidget + ?Sized>(
        &mut self,
    ) -> Option<WidgetMut<'_, W2>> {
        Some(WidgetMut {
            ctx: self.ctx.reborrow_mut(),
            widget: W2::from_dyn_mut(self.widget.as_mut_dyn())?,
        })
    }

    /// Downcasts to `WidgetMut` of concrete Widget type.
    ///
    /// ## Panics
    ///
    /// Panics if the downcast fails, with an error message that shows the
    /// discrepancy between the expected and actual types.
    pub fn downcast<W2: Widget + FromDynWidget + ?Sized>(&mut self) -> WidgetMut<'_, W2> {
        let w1_name = self.widget.type_name();
        match W2::from_dyn_mut(self.widget.as_mut_dyn()) {
            Some(widget) => WidgetMut {
                ctx: self.ctx.reborrow_mut(),
                widget,
            },
            None => {
                panic!(
                    "failed to downcast widget: expected widget of type `{}`, found `{}`",
                    std::any::type_name::<W2>(),
                    w1_name,
                );
            }
        }
    }
}

// TODO - unit tests
