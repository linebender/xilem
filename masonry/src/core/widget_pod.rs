// Copyright 2018 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

use crate::core::{Widget, WidgetId};
use crate::kurbo::Affine;

/// A container for one widget in the hierarchy.
///
/// Generally, container widgets don't contain other widgets directly,
/// but rather contain a `WidgetPod`, which has additional state needed
/// for layout and for the widget to participate in event flow.
///
/// See [container widget tutorial](crate::doc::doc_03_implementing_container_widget) for details.
pub struct WidgetPod<W: ?Sized> {
    id: WidgetId,
    inner: WidgetPodInner<W>,
}

// TODO - This is a simple state machine that lets users create WidgetPods
// without immediate access to the widget arena. It's very inefficient
// and leads to ugly code. The alternative is to force users to create WidgetPods
// through context methods where they already have access to the arena.
// Implementing that requires solving non-trivial design questions.

pub(crate) struct CreateWidget<W: ?Sized> {
    pub(crate) widget: Box<W>,
    pub(crate) transform: Affine,
}

enum WidgetPodInner<W: ?Sized> {
    Create(CreateWidget<W>),
    Inserted,
}

impl<W: Widget> WidgetPod<W> {
    /// Create a new widget pod.
    ///
    /// In a widget hierarchy, each widget is wrapped in a `WidgetPod`
    /// so it can participate in layout and event flow. The process of
    /// adding a child widget to a container should call this method.
    pub fn new(inner: W) -> Self {
        Self::new_with_id(inner, WidgetId::next())
    }

    /// Create a new widget pod with fixed id.
    pub fn new_with_id(inner: W, id: WidgetId) -> Self {
        Self::new_with_id_and_transform(Box::new(inner), id, Affine::IDENTITY)
    }
}

impl<W: Widget + ?Sized> WidgetPod<W> {
    /// Create a new widget pod with a custom transform.
    pub fn new_with_transform(inner: Box<W>, transform: Affine) -> Self {
        Self::new_with_id_and_transform(inner, WidgetId::next(), transform)
    }

    pub fn new_with_id_and_transform(inner: Box<W>, id: WidgetId, transform: Affine) -> Self {
        Self {
            id,
            inner: WidgetPodInner::Create(CreateWidget {
                widget: inner,
                transform,
            }),
        }
    }
    pub(crate) fn incomplete(&self) -> bool {
        matches!(self.inner, WidgetPodInner::Create(_))
    }

    pub(crate) fn take_inner(&mut self) -> Option<CreateWidget<W>> {
        match std::mem::replace(&mut self.inner, WidgetPodInner::Inserted) {
            WidgetPodInner::Create(widget) => Some(widget),
            WidgetPodInner::Inserted => None,
        }
    }

    /// Get the identity of the widget.
    pub fn id(&self) -> WidgetId {
        self.id
    }

    /// Type-erase the contained widget.
    ///
    /// Convert a `WidgetPod` pointing to a widget of a specific concrete type
    /// `WidgetPod` pointing to a `dyn Widget`.
    pub fn erased(self) -> WidgetPod<dyn Widget> {
        let WidgetPodInner::Create(inner) = self.inner else {
            // TODO - Enabling this case isn't impossible anymore.
            // We're keeping it forbidden for now.
            panic!("Cannot box a widget after it has been inserted into the widget graph")
        };
        WidgetPod {
            id: self.id,
            inner: WidgetPodInner::Create(CreateWidget {
                widget: inner.widget.as_box_dyn(),
                transform: inner.transform,
            }),
        }
    }
}
