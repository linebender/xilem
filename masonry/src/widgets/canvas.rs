// Copyright 2025 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

#![warn(missing_docs)]

//! A canvas widget.

use accesskit::{Node, Role};
use smallvec::SmallVec;
use tracing::{trace_span, Span};
use vello::kurbo::Size;
use vello::Scene;

use crate::core::{
    AccessCtx, AccessEvent, BoxConstraints, EventCtx, LayoutCtx, PaintCtx, PointerEvent, QueryCtx,
    RegisterCtx, TextEvent, Update, UpdateCtx, Widget, WidgetId, WidgetMut,
};

/// A widget allowing custom drawing.
pub struct Canvas {
    draw: Box<dyn Fn(&mut Scene, Size) + Send + Sync + 'static>,
    // TODO: pointer events
}

// --- MARK: BUILDERS ---
impl Canvas {
    /// Create a new canvas with the given draw function
    pub fn new(draw: impl Fn(&mut Scene, Size) + Send + Sync + 'static) -> Self {
        Self {
            draw: Box::new(draw),
        }
    }
}

// --- MARK: WIDGETMUT ---
impl Canvas {
    /// Update the draw function
    pub fn update_draw(
        this: &mut WidgetMut<'_, Self>,
        draw: impl Fn(&mut Scene, Size) + Send + Sync + 'static,
    ) {
        this.widget.draw = Box::new(draw);
        this.ctx.request_render();
    }
}

// --- MARK: IMPL WIDGET ---
impl Widget for Canvas {
    fn on_pointer_event(&mut self, _ctx: &mut EventCtx, _event: &PointerEvent) {
        // TODO: ensure coordinates are correct and pass to callback
    }

    fn accepts_pointer_interaction(&self) -> bool {
        true
    }

    fn on_text_event(&mut self, _ctx: &mut EventCtx, _event: &TextEvent) {}

    fn on_access_event(&mut self, _ctx: &mut EventCtx, _event: &AccessEvent) {}

    fn register_children(&mut self, _ctx: &mut RegisterCtx) {}

    fn update(&mut self, _ctx: &mut UpdateCtx, _event: &Update) {}

    fn layout(&mut self, _ctx: &mut LayoutCtx, bc: &BoxConstraints) -> Size {
        // use as much space as possible - caller can size it as necessary
        bc.max()
    }

    fn paint(&mut self, ctx: &mut PaintCtx, scene: &mut Scene) {
        (self.draw)(scene, ctx.size())
    }

    fn accessibility_role(&self) -> Role {
        Role::Canvas
    }

    fn accessibility(&mut self, _ctx: &mut AccessCtx, _node: &mut Node) {
        // TODO: should probably give the caller the opportunity to handle accessibility
    }

    fn children_ids(&self) -> SmallVec<[WidgetId; 16]> {
        SmallVec::new()
    }

    fn make_trace_span(&self, ctx: &QueryCtx<'_>) -> Span {
        trace_span!("Canvas", id = ctx.widget_id().trace())
    }

    fn get_debug_text(&self) -> Option<String> {
        None
    }
}

// --- MARK: TESTS ---
#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use vello::kurbo::{Affine, BezPath, Stroke};
    use vello::peniko::{Color, Fill};

    use super::*;
    use crate::assert_render_snapshot;
    use crate::testing::TestHarness;

    #[test]
    fn simple_canvas() {
        let canvas = Canvas::new(|scene, size| {
            let scale = Affine::scale_non_uniform(size.width, size.height);
            let mut path = BezPath::new();
            path.move_to((0.1, 0.1));
            path.line_to((0.9, 0.9));
            path.line_to((0.9, 0.1));
            path.close_path();
            path = scale * path;
            scene.fill(
                Fill::NonZero,
                Affine::IDENTITY,
                Color::from_rgb8(100, 240, 150),
                None,
                &path,
            );
            scene.stroke(
                &Stroke::new(4.),
                Affine::IDENTITY,
                Color::from_rgb8(200, 140, 50),
                None,
                &path,
            );
        });

        let mut harness = TestHarness::create(canvas);

        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "hello");
    }
}
