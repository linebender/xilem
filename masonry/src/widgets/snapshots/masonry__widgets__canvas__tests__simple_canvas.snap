---
source: masonry/src/widgets/canvas.rs
expression: harness.root_widget()
---
Canvas
